File: flow/script/lib/audio.ts
/**
 * Audio extraction and processing via ffmpeg.
 */

import { BINARIES } from "../config/paths.ts";

const FFMPEG = BINARIES.ffmpeg;

export interface ExtractOptions {
  inputPath: string;
  outputPath: string;
  sampleRate?: number;
  channels?: number;
  bitrate?: string;
  format?: string;
}

/** Extract audio from video using ffmpeg */
export async function extractAudio(options: ExtractOptions): Promise<string> {
  const {
    inputPath,
    outputPath,
    sampleRate = 22050,
    channels = 1,
    bitrate = "64k",
    format = "mp3",
  } = options;

  const cmd = new Deno.Command(FFMPEG, {
    args: [
      "-y",
      "-i", inputPath,
      "-vn",
      "-ar", String(sampleRate),
      "-ac", String(channels),
      "-b:a", bitrate,
      "-f", format,
      outputPath,
    ],
    stdout: "piped",
    stderr: "piped",
  });

  const { code, stderr } = await cmd.output();
  const err = new TextDecoder().decode(stderr);

  if (code !== 0) {
    throw new Error(`ffmpeg exited ${code}: ${err.slice(0, 500)}`);
  }

  // Verify output exists
  try {
    await Deno.stat(outputPath);
  } catch {
    throw new Error(`ffmpeg did not produce output file: ${outputPath}`);
  }

  return outputPath;
}

/** Get audio/video file info via ffprobe */
export async function probeFile(path: string): Promise<{
  duration: number;
  bitrate: number;
  codec: string;
  sampleRate?: number;
}> {
  const ffprobe = FFMPEG.replace("ffmpeg", "ffprobe");
  const cmd = new Deno.Command(ffprobe, {
    args: [
      "-v", "error",
      "-show_entries", "format=duration,bit_rate",
      "-show_entries", "stream=codec_name,sample_rate",
      "-of", "json",
      path,
    ],
    stdout: "piped",
    stderr: "piped",
  });

  const { code, stdout, stderr } = await cmd.output();
  if (code !== 0) {
    throw new Error(`ffprobe failed: ${new TextDecoder().decode(stderr)}`);
  }

  const data = JSON.parse(new TextDecoder().decode(stdout));
  const format = data.format || {};
  const stream = (data.streams || [])[0] || {};

  return {
    duration: parseFloat(format.duration) || 0,
    bitrate: parseInt(format.bit_rate) || 0,
    codec: stream.codec_name || "unknown",
    sampleRate: stream.sample_rate ? parseInt(stream.sample_rate) : undefined,
  };
}

File: flow/script/lib/audio.py
"""
Audio extraction and processing via ffmpeg.
"""

import json
import subprocess
from pathlib import Path
from typing import Optional

from config.paths import BINARIES

FFMPEG = BINARIES["ffmpeg"]
FFPROBE = BINARIES["ffprobe"]


def extract_audio(
    input_path: str,
    output_path: str,
    sample_rate: int = 22050,
    channels: int = 1,
    bitrate: str = "64k",
    audio_format: str = "mp3",
) -> str:
    """Extract audio from video using ffmpeg."""
    cmd = [
        FFMPEG,
        "-y",
        "-i", input_path,
        "-vn",
        "-ar", str(sample_rate),
        "-ac", str(channels),
        "-b:a", bitrate,
        "-f", audio_format,
        output_path,
    ]

    result = subprocess.run(cmd, capture_output=True, text=True, timeout=300)
    if result.returncode != 0:
        raise RuntimeError(f"ffmpeg exited {result.returncode}: {result.stderr[:500]}")

    if not Path(output_path).exists():
        raise FileNotFoundError(f"ffmpeg did not produce output: {output_path}")

    return output_path


def probe_file(path: str) -> dict[str, any]:
    """Get media file metadata via ffprobe."""
    cmd = [
        FFPROBE,
        "-v", "error",
        "-show_entries", "format=duration,bit_rate",
        "-show_entries", "stream=codec_name,sample_rate",
        "-of", "json",
        path,
    ]

    result = subprocess.run(cmd, capture_output=True, text=True, timeout=60)
    if result.returncode != 0:
        raise RuntimeError(f"ffprobe failed: {result.stderr}")

    data = json.loads(result.stdout)
    fmt = data.get("format", {})
    stream = (data.get("streams") or [{}])[0]

    return {
        "duration": float(fmt.get("duration", 0)),
        "bitrate": int(fmt.get("bit_rate", 0)),
        "codec": stream.get("codec_name", "unknown"),
        "sample_rate": int(stream.get("sample_rate", 0)) if stream.get("sample_rate") else None,
    }
